### ISSUE TITLE
Are non-manifold meshes (e.g. a "T-junction" in a shell geometry) unsupported in dolfinx?

### ISSUE THREAD

**Issue by austinherrema** [FEniCS/dolfinx#1465]
I am hoping to use `dolfinx` to perform structural analysis of a relatively complex shell geometry. At the moment I am hoping to use `dolfinx` as opposed to legacy `dolfin` due to the better support of quad elements as discussed [here](https://bitbucket.org/fenics-project/dolfin/issues/1089/quadrilateral-mesh-reordering-error).

I made a simple T-junction geometry (.3dmf attached with .txt extension)

![image](https://user-images.githubusercontent.com/29387274/112530782-78f15b80-8d6c-11eb-8661-23e1bac83592.png)

and am trying to load it into `dolfinx` using:

``` python
import dolfinx
from dolfinx import io
from dolfinx.io import XDMFFile
from mpi4py import MPI

infile = XDMFFile(MPI.COMM_WORLD, "tTopology.xdmf", "r", encoding=dolfinx.cpp.io.XDMFFile.Encoding.ASCII)
mesh = infile.read_mesh(name="Grid")
infile.close()
```

This gives the following error:
```
2021-03-23 15:08:14.799 ( 0.291s) [main ] graphbuild.cpp:93 ERR| Found three identical facets in mesh (local)
Traceback (most recent call last):
File “xdmfToPvd.py”, line 11, in
mesh = infile.read_mesh(name=“Grid”)
File “/usr/local/dolfinx-real/lib/python3.8/dist-packages/dolfinx/io.py”, line 59, in read_mesh
mesh = cpp.mesh.create_mesh(self.comm(), cpp.graph.AdjacencyList_int64(cells),
RuntimeError: Inconsistent mesh data in GraphBuilder: three identical facets
```

Error is clearly stemming from [`graphbuild.cpp`](https://github.com/FEniCS/dolfinx/blob/4cdc251c325adbf884150aa00914da5eec1c2f40/cpp/dolfinx/mesh/graphbuild.cpp). Is this completely expected in this case and, if so, are there any known workarounds? I can imagine, for example, dividing the mesh into multiple manifold meshes and then handling the geometry coupling manually, but this would obviously be less straightforward,

[tTopology.txt](https://github.com/FEniCS/dolfinx/files/6207313/tTopology.txt)

**Comment by chrisrichardson** [comment]
It's a check, which can be disabled... I'll take a look.

**Comment by garth-wells** [comment]
@chrisrichardson presumably the test can be changed to check the topological dimension so that it still raises an exception if more than two facets of topological dimension 2 are connected?

**Comment by chrisrichardson** [comment]
To do it properly, we should know the geometric dimension. In 2D on a plane, it is still an error to have more than two cells connected to a facet.

**Comment by RuruX** [comment]
Hi @chrisrichardson, 

Our team is working on the structural analysis of an airplane model with shell elements, where there is some sort of "t-junction geometry" in the mesh. We saw the same error as @austinherrema when openning the mesh with DOLFIN-X. And we were able to fix this by changing the check to only raise the error when topological dimension equals to three in ['graphbuild.cpp'](https://github.com/FEniCS/dolfinx/blob/43b9dab48c574fe267e8349a84b7a86eb37ed4f1/cpp/dolfinx/mesh/graphbuild.cpp#L166).

```
      // FIXME: This may not strictly be an error if tdim != gdim
      if (eq_count == 2 && tdim == 3)
        throw std::runtime_error("Same facet in more than two cells");
    }
```
I totally see your point that it could be more nicely done if we have the geometric dimension. But for now, we cannot find it in the "graphbuild.cpp" yet. And making the "t-junction" supported by DOLFIN-X with the approach above would be more beneficial for people having a model like ours, than problematic for those with a bad mesh, as it's not DOLFIN-X's responsibility to provide a reasonable mesh but the users'. Please let me know what you think!

Thanks,
Ru

**Comment by chrisrichardson** [comment]
Clearly this is causing problems for a few people - I'll make a change.

**Comment by garth-wells** [comment]
Fixed in #1584.

**Comment by jorgensd** [comment]
This should definitely be solved with @schnellerhase's #3733.
Here is an MWE of reading in
```
<Xdmf Version="3.0">
  <Domain>
    <Grid Name="Grid">
      <Geometry GeometryType="XYZ">
        <DataItem DataType="Float" Dimensions="8 3" Format="XML" Precision="8">
            0.000000000000000 0.000000000000000 0.000000000000000
            1.000000000000000 0.000000000000000 0.000000000000000
            1.000000000000000 0.000000000000000 1.000000000000000
            0.000000000000000 0.000000000000000 1.000000000000000
            -1.000000000000000 0.000000000000000 1.000000000000000
            -1.000000000000000 0.000000000000000 0.000000000000000
            0.000000000000000 1.000000000000000 0.000000000000000
            0.000000000000000 1.000000000000000 1.000000000000000
        </DataItem>
      </Geometry>
      <Topology NodesPerElement="4" NumberOfElements="4" TopologyType="quadrilateral">
        <DataItem DataType="Int" Dimensions="3 4" Format="XML" Precision="8">
            0 1 2 3
            0 3 4 5
            0 3 7 6
        </DataItem>
      </Topology>
    </Grid>
  </Domain>
</Xdmf>

```

```python
from mpi4py import MPI
import dolfinx
import numpy as np
import ufl
import basix.ufl
from dolfinx.cpp.mesh import create_cell_partitioner
mpi_comm = MPI.COMM_WORLD

xdmf = dolfinx.io.XDMFFile(mpi_comm, "joint.xdmf", "r", encoding=dolfinx.io.XDMFFile.Encoding.ASCII)

x = xdmf.read_geometry_data(name="Grid")
topo = xdmf.read_topology_data(name="Grid")
cell_shape, cell_degree = xdmf.read_cell_type(name="Grid")
domain = ufl.Mesh(basix.ufl.element("Lagrange", cell_shape.name, cell_degree, shape=(3,)))


part = create_cell_partitioner(dolfinx.mesh.GhostMode.shared_facet, 3)
new_mesh = dolfinx.mesh.create_mesh(mpi_comm, topo, domain, x, part)

cell_map = new_mesh.topology.index_map(new_mesh.topology.dim)
vertex_map = new_mesh.topology.index_map(0)
print(f"Number of cells: {cell_map.size_global}, Number of vertices: {vertex_map.size_global}")
```
